R/framework.R

```diff
@@ -63,25 +63,78 @@ trading_dates(start, end, calendar=holidayNYSE) %as% {
   as.Date(dates)
 }
 
+trading_dates(start, obs, calendar=holidayNYSE) %::% a:numeric:Function:.
+trading_dates(start, obs, calendar=holidayNYSE) %when% { obs==0 } %as% {
+  # more reasonable to return NULL instead of as.Date(NA) here.
+  # And NULL cannot be coerced into class Date,
+  # so set return type to . here
+  NULL
+}
+
 trading_dates(start, obs, calendar=holidayNYSE) %::% a:numeric:Function:Date
-trading_dates(start, obs, calendar=holidayNYSE) %when% { obs>=0 } %as% {
+trading_dates(start, obs, calendar=holidayNYSE) %when% { obs>0 } %as% {
   start <- as.Date(start)
-  # This is to get enough dates to account for holidays and weekends
-  shimmed <- ceiling(obs * 2)
-  dates <- timeSequence(from=start, length.out=shimmed)
-  dates <- as.Date(dates[isBizday(dates, holidays=calendar(unique(year(dates))))])
-  dates <- dates[dates >= start]
-  dates <- dates[1:obs]
+  # if use other numbers here, 
+  # make sure step.size * max.its can always cover obs trading days
+  step.size <- 50 
+  # cannot use infinity here, memory will blow up
+  max.its <- ceiling(obs/step.size)*2 
+  holidays_save <- list() 
+  dates <- callCC(function(ext) fold(1:max.its, function(i, dates){
+      if (length(dates) >= obs) ext(dates[1:obs])
```

Comment:
Three issues:
1. This conditional is unnecessary if you use `start` as initial value for `fold` instead of `NULL`;
2. It seems that this function is duplicated below, meaning it should be pulled out as a separate function;
3. the function itself is rather inefficient. It would be faster to construct the sequence, find holidays, add extra days to account for holidays, repeat.